api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/activation/ELU.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2021 JetBrains s.r.o. and Kotlin Deep Learning project contributors. All Rights Reserved.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlinx.dl.api.core.layer.activation
+
+import org.tensorflow.Operand
+import org.tensorflow.op.Ops
+
+/**
+ * Exponential Unit activation function.
+ *
+ * It follows:
+ * ```
+ * f(x) = x,                    if x > 0
+ * f(x) = alpha * (exp(x) - 1), if x <= 0
+ * ```
+ *
+ * In contrast to ReLU it has negative values which push the mean of
+ * the activation closer to zero which enable faster learning as they
+ * bring the gradient to the natural gradient.
+ *
+ * @property [alpha] Hyperparameter that controls the value to which
+ * an ELU saturates for negative net inputs. Should be > 0.
+ * @constructor Creates [ELU] object.
+ * @since 0.3
+ */
+public class ELU(
+    public val alpha: Float = 1.0f,
+    name: String = ""
+) : Activation(name) {
+    init {
+        require(alpha > 0.0f) { "Alpha $alpha should be > 0.0." }
+    }
+
+    override fun forward(
+        tf: Ops,
+        input: Operand<Float>
+    ): Operand<Float> = when(alpha) {
+        1.0f -> tf.nn.elu(input)
+        else -> {
+            val greaterThanZero = tf.math.greater(input, tf.constant(0.0f))
+            val scaledActivation = tf.math.mul(tf.constant(alpha), tf.nn.elu(input))
+            tf.where3(greaterThanZero, input, scaledActivation)
```

Comment:
do we have a chance to avoid where op here (I'm not sure, but maybe the gradients are missed for this and this layer should be marked with NoGradients interface and should be frozen by default) Don't mind if you have no able to rewrite it, leave it this way, but mark with NoGradient interface

Reply:
imo we can mark the AbstractActivationLayer as not trainable and so all they child classes as it's seems to be more common for the activations not to be trainable (and as I correctly understand it will not be necessary to use the NoGradient interface then and even harder to override in some subclasses which could have been trainable).

Btw interface here doesn't look like the best approach, maybe it should be redesigned with annotation (so it could mark only specific class and not its children)?

Reply:
At this time, NoGradient is a marker interface, and it's very simple to verification. There is no need to mark the AbstractActivationLayer with this interface because it's not true for the ReLU layer, which could participate in backward computation.

It looks like NoGradient is an unclear name for the interface. Still, the idea is related to weights (activation is just a function applied to the previous layer's outputs), but the problem is in having the GradientOp for this TF Op. For many TF Ops gradients (read here: a building block for backward computing is missed).

I suggest don't concentrate on this now, but just marked the class with this interface. Later I'll do a double-check for used TF ops